pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -64,18 +96,138 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
+
+	// trigger the event if the message type is registered
+	ok, err := s.sendEvent(msg, thid, next)
+	if err != nil {
+		return err
+	}
+
+	// if no events are trigger continue the execution
+	if !ok {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// startCallbackListener starts the callback listeners. The protocol service will trigger registered events on incoming
+// message types along with a callback channel. These callback channels are received here.
+func (s *Service) startCallbackListener() {
+	go func() {
+		for {
+			select {
+			case msg := <-s.callback[connectionInvite]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionRequest]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionResponse]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionAck]:
+				s.processCallback(msg)
+			}
+		}
+	}()
+}
+
+// processCallback processes the callback events.
+func (s *Service) processCallback(didCommCallback event.DIDCommCallback) {
+	// TODO what to do in case of any failures with callback processing ? - currently, logging the error messages and
+	//  ignoring the flow.
+	if didCommCallback.Err != nil {
+		logger.Errorf("inbound event processing failed: %v", didCommCallback.Err)
+		return
+	}
+
+	// fetch the record
+	jsonDoc, err := s.store.Get(didCommCallback.ID)
+	if err != nil {
+		logger.Errorf("document for the id doesn't exists in the database %v", err)
+		return
+	}
+
+	document := &message{}
+	err = json.Unmarshal(jsonDoc, document)
+	if err != nil {
+		logger.Errorf("JSON marshalling failed: %v", err)
+		return
+	}
+
+	// continue the processing
+	err = s.handle(document)
+	if err != nil {
+		logger.Errorf("processing of the message failed: %v", err)
+	}
+}
+
+func (s *Service) handle(msg *message) error {
+	nextState, err := stateFromName(msg.NextStateName)
+	if err != nil {
+		return fmt.Errorf("invalid state name: %w", err)
+	}
+
 	// TODO execute actions for the current state *and the next (if any)*. Eg. if we receive an
 	//      invitation, execute actions for 'invited' state and then execute actions for
 	//      'requested' state. Implement the relevant actions for each state in state.go.
-	err = s.update(thid, next)
+	err = s.update(msg.ThreadID, nextState)
 	if err != nil {
 		return err
 	}
 	// TODO call post-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
 	return nil
 }
 
+// Register events on DID Exchange protocol.
+func (s *Service) Register(msgType string, ch chan<- event.DIDCommEvent) (chan<- event.DIDCommCallback, error) {
```

Comment:
Looks like we need to export the msgTypes as well.

Reply:
Valid point. [Updated](https://github.com/hyperledger/aries-framework-go/compare/35ffd11efb21de8bceb7b468e53a833a2b1a8f82..e20d32d8200881f1bfdc2f36aad1988fa5a78e93#diff-f761c91e753eba1c42a68076df06ce12R34).